### PR TITLE
[MI-2656] Add test cases for api.go and client.go

### PR DESCRIPTION
**Pull request by ayusht2810**

### CODE REVIEW

**Review by raghavaggarwal2308 — APPROVED** [review]

File: server/plugin/api_test.go [hunk 1]
```diff
@@ -1060,7 +1060,7 @@ func TestCreateSubscription(t *testing.T) {
 		ExpectedStatusCode   int
 		ExpectedErrorMessage string
 	}{
-		"success": {
+		"success with subscription type record": {
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:1063]: we can do like this maybe ?
```suggestion
		"success with subscription type: record": {
```

File: server/plugin/api_test.go [hunk 2]
```diff
@@ -1088,6 +1088,67 @@ func TestCreateSubscription(t *testing.T) {
 			},
 			ExpectedStatusCode: http.StatusCreated,
 		},
+		"success with subscription type bulk": {
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:1091]: same ?

File: server/plugin/api_test.go [hunk 3]
```diff
@@ -1088,6 +1088,67 @@ func TestCreateSubscription(t *testing.T) {
...
+
+				monkey.PatchInstanceMethod(reflect.TypeOf(p), "HasChannelPermissions", func(_ *Plugin, _, _ string) (int, error) {
+					return http.StatusOK, nil
+				})
+			},
+			ExpectedStatusCode: http.StatusCreated,
+		},
+		"error with subscription type bulk": {
+			RequestBody: fmt.Sprintf(`{
+				"user_id": "%s",
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:1126]: create a mock function for this request body with param record or bulk subscription and use that here

File: server/plugin/util_test.go [hunk 4]
```diff
@@ -66,6 +67,54 @@ func TestParseSubscriptionsToCommandResponse(t *testing.T) {
...
+			},
+		},
+	} {
+		t.Run(testCase.description, func(t *testing.T) {
+			p, api := setupTestPlugin(&plugintest.API{}, nil)
+			client := setupPluginForCheckOAuthMiddleware(p, t)
+			testCase.setupClient(client)
+			testCase.setupAPI(api)
+			defer api.AssertExpectations(t)
+			p.GetFiltersFromServiceNow(testCase.subscription, client, nil, true)
```
Comment (Kshitij-Katiyar) [inline server/plugin/util_test.go:113]: what are you checking in this test case ?

File: server/plugin/command_test.go [hunk 5]
```diff
@@ -802,6 +847,57 @@ func TestHandleEditSubscription(t *testing.T) {
...
+		UserId: testutils.GetID(),
+	}
+
+	t.Run("HandleCreateRequest: Success", func(t *testing.T) {
+		defer mockAPI.AssertExpectations(t)
+		assert := assert.New(t)
+		mockAPI.On("PublishWebSocketEvent", mock.AnythingOfType("string"), mock.Anything, mock.AnythingOfType("*model.WebsocketBroadcast")).Return()
+		p.SetAPI(mockAPI)
+		resp := p.handleCreateRequest(args)
+		assert.EqualValues("", resp)
```
Comment (Kshitij-Katiyar) [inline server/plugin/command_test.go:880]: here resp is going to be empty everytime or just for the successful and non empty for error case ?

File: server/plugin/command_test.go [hunk 6]
```diff
@@ -445,6 +445,34 @@ func TestHandleCreate(t *testing.T) {
...
+func TestHandleRecords(t *testing.T) {
+	p := Plugin{}
+	args := &model.CommandArgs{
+		UserId: testutils.GetID(),
+	}
+	for _, testCase := range []struct {
+		description      string
+		params           []string
+		expectedResponse string
+	}{
```
Comment (Kshitij-Katiyar) [inline server/plugin/command_test.go:457]: any successful test case here ?

File: server/plugin/api_test.go [hunk 7]
```diff
@@ -1913,6 +1893,98 @@ func TestAPICreateIncident(t *testing.T) {
...
+				})
+
+				var p *Plugin
+				monkey.PatchInstanceMethod(reflect.TypeOf(p), "HasChannelPermissions", func(_ *Plugin, _, _ string, _ bool) (int, string, error) {
+					return http.StatusOK, model.CHANNEL_DIRECT, nil
+				})
+			},
+			ExpectedStatusCode: http.StatusOK,
+		},
+		"error getting direct channel": {
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:1925]: add the

File: server/plugin/command_test.go (hunk 5 above)
Comment (ayusht2810) [reply server/plugin/command_test.go:965]: yes it is going to be empty every time

Comment (Kshitij-Katiyar) [reply server/plugin/command_test.go:965]: then why are we checking this anyways if it going to be empty everytime ?

File: server/plugin/util_test.go (hunk 4 above)
Comment (Kshitij-Katiyar) [reply server/plugin/util_test.go:113]: ok but we usually don't consider AssertExpectations as a test case check. There are many reasons for it:-
1) It just checks whether methods are added using API.On is called and returned as expected. But it doesn't check how many times that method is called. What is the order of the method called etc.
That's why if a function is not returning anything we check how many times a specific method is called or not.

File: server/plugin/command_test.go (hunk 5 above)
Comment (ayusht2810) [reply server/plugin/command_test.go]: discussed

**Review by Kshitij-Katiyar — APPROVED** [review]

**Pull request merged**